compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -1,38 +1,45 @@
-//! Generalized type folding mechanism. The setup is a bit convoluted
-//! but allows for convenient usage. Let T be an instance of some
-//! "foldable type" (one which implements `TypeFoldable`) and F be an
-//! instance of a "folder" (a type which implements `TypeFolder`). Then
-//! the setup is intended to be:
+//! A generalized traversal mechanism for complex data structures that contain
+//! type information.
 //!
-//!     T.fold_with(F) --calls--> F.fold_T(T) --calls--> T.super_fold_with(F)
```

Comment:
I think it would be nice to have a small section detailing what calls what in a small example, like say `struct T(Ty, U)` where `T: TypeFoldable` and `U: TypeFoldable`:

`T.fold_with(visitor) -> T.super_fold_with(visitor) -> U.fold_with(visitor) and Ty.fold_with(visitor) -> visitor.fold_ty(Ty) --often-> Ty.super_fold_with(visitor)`

pretty sure you could figure this out from the docs on `try_super_fold_with` and `try_fold_with` but i still think it'd be nice to have as part of module docs :)